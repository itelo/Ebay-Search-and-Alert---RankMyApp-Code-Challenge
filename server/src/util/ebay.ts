import fetch from "node-fetch";

function findItemsByKeywords(keyword: string) {
  return fetch(
    `https://svcs.sandbox.ebay.com/services/search/FindingService/v1?SERVICE-NAME=FindingService&OPERATION-NAME=findItemsByKeywords&SERVICE-VERSION=1.12.0&SECURITY-APPNAME=${
      process.env.EBAY_APP_ID
    }&RESPONSE-DATA-FORMAT=JSON&REST-PAYLOAD&GLOBAL-ID=EBAY-US&keywords=${keyword}&paginationInput.entriesPerPage=3&paginationInput.pageNumber=1&SortOrderType=PricePlusShippingLowest`
  ).then(r => r.json());
}

type Item = {
  _id: string;
  price: string;
  itemURL: string;
  title: string;
  imageURL: string;
};

export const findItemsBySearchPhrase = (searchPhrase: string) => {
  return new Promise<Item[]>((resolve, reject) => {
    findItemsByKeywords(searchPhrase).then(
      (data: any) => {
        // console.log(data.findItemsByKeywordsResponse[0].searchResult[0]);
        let result = [] as Item[];
        if (
          data &&
          data.findItemsByKeywordsResponse &&
          data.findItemsByKeywordsResponse[0] &&
          data.findItemsByKeywordsResponse[0].searchResult &&
          data.findItemsByKeywordsResponse[0].searchResult[0] &&
          data.findItemsByKeywordsResponse[0].searchResult[0].item
        ) {
          result = data.findItemsByKeywordsResponse[0].searchResult[0].item.map(
            (item: any) => {
              return {
                price: item.sellingStatus[0].currentPrice[0].__value__,
                itemURL: item.viewItemURL[0],
                _id: item.itemId[0],
                title: item.title[0],
                imageURL:
                  item.galleryURL && item.galleryURL[0]
                    ? item.galleryURL && item.galleryURL[0]
                    : undefined
              };
            }
          ) as Item[];
        }

        resolve(result);
      },
      (error: any) => {
        reject(error);
      }
    );
  });
};
