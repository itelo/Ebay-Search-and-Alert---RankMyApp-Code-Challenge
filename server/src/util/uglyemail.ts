type Item = {
  price: string;
  itemURL: string;
  title: string;
  imageURL: string;
};

export const uglyEbayItem = (item: Item) => {
  const { price, itemURL, title, imageURL } = item;

  const image = imageURL
    ? `<td><img src="${imageURL}" alt="" width="96" height="96" style="display: block;" /></td>`
    : "";
  return `
    <tr>
      <td width="260" valign="top">
        <table border="0" cellpadding="0" cellspacing="0" width="100%">
          <tr style="cursor:pointer"
            onclick="window.open('${itemURL}')">
            ${image}
            <td style="padding: 0 0 0 0;">
              <table border="0" cellpadding="0" cellspacing="0">
                <tr>
                  <td style="padding: 0 0 0 0;">
                    ${title}
                  </td>
                <tr>
                  <td style="padding: 0 0 0 0;">
                    US $${price}
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        </table>
      </td>
    </tr>`;
};
