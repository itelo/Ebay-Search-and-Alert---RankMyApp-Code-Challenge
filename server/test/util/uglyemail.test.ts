import { uglyEbayItem } from "../../src/util/uglyemail";

describe("uglyEbayItem", () => {
  it("should return table without image", () => {
    const ebayItem = {
      price: "12,00",
      itemURL: "http://www.some-url.com",
      title: "some product"
    };
    const table = uglyEbayItem(ebayItem);
    expect(table).toMatchSnapshot();
  });

  it("should return table without image", () => {
    const ebayItem = {
      price: "12,00",
      itemURL: "http://www.some-url.com",
      title: "some product",
      imageURL: "http://www.some-image-url.com"
    };
    const table = uglyEbayItem(ebayItem);
    expect(table).toMatchSnapshot();
  });
});
