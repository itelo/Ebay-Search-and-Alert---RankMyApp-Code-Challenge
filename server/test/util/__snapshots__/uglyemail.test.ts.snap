// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uglyEbayItem should return table without image 1`] = `
"
    <tr>
      <td width=\\"260\\" valign=\\"top\\">
        <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\">
          <tr style=\\"cursor:pointer\\"
            onclick=\\"window.open('http://www.some-url.com')\\">
            
            <td style=\\"padding: 0 0 0 0;\\">
              <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\">
                <tr>
                  <td style=\\"padding: 0 0 0 0;\\">
                    some product
                  </td>
                <tr>
                  <td style=\\"padding: 0 0 0 0;\\">
                    US $12,00
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        </table>
      </td>
    </tr>"
`;

exports[`uglyEbayItem should return table without image 2`] = `
"
    <tr>
      <td width=\\"260\\" valign=\\"top\\">
        <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\">
          <tr style=\\"cursor:pointer\\"
            onclick=\\"window.open('http://www.some-url.com')\\">
            <td><img src=\\"http://www.some-image-url.com\\" alt=\\"\\" width=\\"96\\" height=\\"96\\" style=\\"display: block;\\" /></td>
            <td style=\\"padding: 0 0 0 0;\\">
              <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\">
                <tr>
                  <td style=\\"padding: 0 0 0 0;\\">
                    some product
                  </td>
                <tr>
                  <td style=\\"padding: 0 0 0 0;\\">
                    US $12,00
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        </table>
      </td>
    </tr>"
`;
